### ISSUE TITLE
Bridge defined on Raspberry Pi reconnect issue

### ISSUE THREAD

**Issue by Bas44** [eclipse/mosquitto#1301]
**Goal:**
* Deliver sensor data from a sensor device in the field to a server. No data loss may occur

**Situation:**
* 2 machines, a sensor (Raspberry Pi, Raspbian version 9 (stretch)) and a server (now a test machine running Windows 7 pro, server is Windows server 2016) 
* 2 brokers: Mosquitto 1.6.2 and 1.6.2 installed (also tried with 1.5.6, same issue)
* Client applications deliver data to the broker on the sensor
* The broker on the sensor should forward messages to the broker on the server, so a bridge is configured on the sensor
* Persistent queues
* Named users
* QoS 1

**Config on server:**
`listener 1883
listener 8883
log_dest file D:\Large data sets\Mosquitto\Persistent store\mosquitto.log
log_type debug

max_queued_messages 10000
autosave_interval 5
autosave_on_changes true
persistence true
persistence_file mosquitto.db
persistence_location D:\Large data sets\Mosquitto\Persistent store\
`

**Config on Sensor:**
`listener 1883
listener 8883
log_type debug

connection testSensor01
address *********** 1883
cleansession false
bridge_attempt_unsubscribe false
topic test out 1

max_queued_messages 10000
autosave_interval 5
autosave_on_changes true
persistence true
persistence_file mosquitto.db
persistence_location /usr/local/SensorPlatform/Mosquitto/PersistentStore/
`
(the *** denote the server IP)

**Issue:**
* Upon startup of the sensor, the broker on the sensor properly connects to the local queue and to the server
* At this point, messages are forwarded to the server
* Then, the broker on the server is stopped, halting the bridge
* Extra messages are sent to the sensor broker
* Upon restart of the server broker, the the bridge is not resumed from the sensor
* If Mosquitto on the sensor is restarted, the bridge works again and the messages not sent so far are delivered to the server.

**Mosquitto log on the sensor:**
`
1559295617: Bridge raspberrypi.testSensor01 sending CONNECT
1559295617: Received CONNACK on connection local.raspberrypi.testSensor01.
1559295617: Sending PUBLISH to local.raspberrypi.testSensor01 (d0, q1, r0, m163, 'test', ... (2 bytes))
1559295617: Sending PUBLISH to local.raspberrypi.testSensor01 (d0, q1, r0, m164, 'test', ... (2 bytes))
1559295617: Sending PUBLISH to local.raspberrypi.testSensor01 (d0, q1, r0, m165, 'test', ... (2 bytes))
.........
1559295431: Sending CONNACK to sender3 (0, 0)
1559295431: Received PUBLISH from sender3 (d0, q1, r0, m1, 'test', ... (2 bytes))
1559295431: Sending PUBACK to sender3 (m1, rc0)
1559295431: Received DISCONNECT from sender3
1559295433: Bridge local.raspberrypi.testSensor01 doing local SUBSCRIBE on topic test
1559295441: Bridge local.raspberrypi.testSensor01 doing local SUBSCRIBE on topic test
1559295442: Bridge local.raspberrypi.testSensor01 doing local SUBSCRIBE on topic test
1559295449: Bridge local.raspberrypi.testSensor01 doing local SUBSCRIBE on topic test
`
On top is the start of Mosquitto. It sends the persistent messages stored from the previous time it ran to the server. After the break there is one message sent. After that I stopped Mosquitto on the server and added some more messages on the server. It keeps repeating the 'doing local SUBSCRIBE' forever and the bridge is not restored. As indicated above the messages do arrive once Mosquitto on the sensor is restarted.

**Messages in the field are sent:**
`mosquitto_pub -h 127.0.0.1 -t test -m FF -q 1 -c -i sender3
`

**A listener on the server is:**
`mosquitto_sub -h 127.0.0.1 -t test -q 1 -c -i subscriber1`

**Note:**
The odd thing is that if the bridge is configured on the Windows machine and is configured in the 'in' direction, it works as expected. The server broker is not stopped in this scenario though, the network connection is interrupted instead.

**Comment by ralight** [comment]
Are you able to try out the `fixes` branch? I've just committed some fixes that are related to this sort of issue.

**Comment by Bas44** [comment]
I assume that would involve compiling Mosquitto? I did not see an option to show branches using the package manager (on the Raspberry). Unfortunately I have no experience in compiling Mosquitto on the Pi.

**Comment by ralight** [comment]
Yes, I'm afraid that is the case. It's not too tricky though.

```
sudo apt install git make gcc libssl-dev libwebsockets-dev 
git clone https://github.com/eclipse/mosquitto
cd mosquitto/src
git checkout fixes
make 
sudo make prefix=/usr install
```

Let me know if you have any problems.

**Comment by Bas44** [comment]
Thanks for the clear instructions, I was able to build and install the fixes branch. FYI: there was one warning. Please see the image:

![2019-06-03 Mosquitto build error](https://user-images.githubusercontent.com/51169816/58803461-da808d00-860f-11e9-9baa-a618e0272233.png)

This version indeed fixes the reconnect problem. It seems that the reconnect is a bit faster for the situation where the server broker is restarted (~ 7 seconds) compared to the situation where the network cable to the Pi is unplugged. That takes about 30 seconds, but it may also be related to a delayed reconnect attempt (I think I've read something like that in the manual). The server broker is not restarted in that case.

I've also tested the situation where the server broker is off and the Pi is starting up (automatically starting Mosquitto). Starting the server broker after some minutes results in a working bridge, also after about 30 - 40 seconds.

**Comment by Bas44** [comment]
Is the code currently in the fixes branch expected to be released as part of the next Mosquitto version? Or is it planned for later releases?
In case I would need to do a rollout of the separately compiled package, is it then suffitient to copy only the executable or do I also need to take other files into account?

**Comment by ralight** [comment]
It will become 1.6.3 next week. I'm just back from holiday and will be back properly on Mosquitto on Tuesday.

**Comment by ralight** [comment]
And thanks for the feedback.

**Comment by Bas44** [comment]
Thanks for your work on Mosquitto

**Comment by ckrey** [comment]
I think this can be closed as fixed in 1.6.3

**Comment by Bas44** [comment]
If I recall it correctly it indeed worked fine in later versions
